pkg/secretpath/path.go

```diff
@@ -0,0 +1,226 @@
+func HasVersion(path string) bool {
+	return hasVersion(path)
+}
+
+// hasVersion returns true when a version suffix is specified in the path.
+func hasVersion(path string) bool {
+	i := strings.LastIndex(path, versionSep)
+	return i >= 0 && i < len(path)-1
+}
+
+// Version returns the version number suffix of a path, returning -1
+// when :latest or negative version numbers are given. If no version
+// suffix is set or it cannot be parsed to an integer, 0 is returned.
+func Version(path string) int {
+	if !hasVersion(path) {
+		return 0
+	}
+
+	if strings.HasSuffix(path, latestSuffix) {
+		return -1
+	}
+
+	i := strings.LastIndex(path, versionSep)
+	version, err := strconv.Atoi(path[i+1:])
+	if err != nil {
+		return 0
+	}
+
+	if version < 0 {
+		return -1
+	}
+
+	return version
+}
+
+// AddVersion adds a version suffix to a given path, removing trailing
+// separators if necessary. If the path already contains a version suffix,
+// it is replaced by the given version number. Negative version numbers
+// are converted to ":latest".
+func AddVersion(path string, version int) string {
+	path = strings.Trim(path, elemSep)
+
+	suffix := latestSuffix
+	if version > 0 {
+		suffix = versionSep + strconv.Itoa(version)
+	}
+
+	i := strings.LastIndex(path, versionSep)
+	if i < 0 {
+		return path + suffix
+	}
+
+	return Clean(path[:i] + suffix)
+}
+
+// Base returns the last element of a path. Trailing separators and version
+// numbers are removed.
+func Base(path string) string {
+	path = Clean(path)
+
+	for i := len(path) - 1; i >= 0; i-- {
+		if path[i] == elemSepByte {
+			return path[i+1:]
+		}
+
+		if path[i] == versionSepByte {
+			path = path[:i]
+		}
+	}
+	return path
+}
+
+// Parent returns all but the last element of a path, removing trailing separators.
+// If a path contains only one element, it returns an empty string.
+func Parent(path string) string {
+	path = Clean(path)
+
+	for i := len(path) - 1; i >= 0; i-- {
+		if path[i] == elemSepByte {
+			return path[:i]
+		}
+	}
+
+	return ""
+}
+
+// Repo returns the first two elements of a path, removing trailing separators.
+// If a path contains less than two elements, it returns the empty string.
+func Repo(path string) string {
+	path = Clean(path)
+
+	count := 0
+	for i := 0; i < len(path); i++ {
+		if path[i] == elemSepByte {
+			count++
+		}
+
+		if count == 2 {
+			return path[:i]
+		}
+	}
+
+	if count == 1 {
+		return path
+	}
+	return ""
+}
+
+// Namespace returns the first element of a path, removing trailing separators.
+func Namespace(path string) string {
+	path = Clean(path)
+	if len(path) == 0 {
+		return ""
+	}
+
+	for i := 0; i < len(path); i++ {
+		if path[i] == elemSepByte {
+			return path[:i]
+		}
+	}
+
+	return path
+}
+
+// Clean returns the shortest path name equivalent to the given path
+// by lexical processing. It removes trailing and multiple separator
+// elements. Version suffixes are not removed.
+func Clean(path string) string {
+	if len(path) == 0 {
```

Comment:
This check is redundant; For the empty string, `strings.SplitN` will return an array containing only the empty string; The loop will have 0 iterations, so the final return will be an empty byte array with the empty string concatenated.

Reply:
It's not required, but a bit more performant I'd say. It's a small and very simple optimization to reduce the number of allocations. Seeing as the function get's called for almost every library call, it's worth the 'trade-off' IMO.

Reply:
This check indeed reduces the number of memory allocations and CPU cycles needed when `Clean` is called with an empty path.

It also increases the CPU cycles needed when `Clean` is called with any other path.

And, as you said, the function gets called for almost every library call.

---

I'd say these library calls are useful when called on non-empty secret paths. Therefore, it might be better to optimize for this useful usage instead of an edge-case empty string usecase.